File: java/dsa/algorithms/strings/is-subsequence.java
class Solution {
    public boolean isSubsequence(String s, String t) {
        int i = 0, j = 0;
        while (i < s.length() && j < t.length()) {
            // 1️⃣ Compare: Do characters match?
            if (s.charAt(i) == t.charAt(j)) {
                // 2️⃣ Match! Consume from subsequence
                i++;
            }
            // 3️⃣ Always advance through main string
            j++;
        }
        return i == s.length();
    }
}

File: python/dsa/algorithms/strings/is-subsequence.py
class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        i, j = 0, 0
        while i < len(s) and j < len(t):
            # 1️⃣ Compare: Do characters match?
            if s[i] == t[j]:
                # 2️⃣ Match! Consume from subsequence
                i += 1
            # 3️⃣ Always advance through main string
            j += 1
        return i == len(s)